core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -0,0 +1,250 @@
+    }
+  }
+
+  private void initChannel() throws IOException {
+    if (outputFileStream == null) {
+      outputFileStream = new FileOutputStream(outputTempFile, true);
+    }
+    if (outputFileChannel == null) {
+      outputFileChannel = outputFileStream.getChannel();
+    }
+  }
+
+  private class DefaultShufflePartitionWriter implements ShufflePartitionWriter {
+
+    private final int partitionId;
+    private PartitionWriterStream stream = null;
+    private PartitionWriterChannel channel = null;
+
+    private DefaultShufflePartitionWriter(int partitionId) {
+      this.partitionId = partitionId;
+    }
+
+    @Override
+    public OutputStream openStream() throws IOException {
+      stream = new PartitionWriterStream();
+      return stream;
+    }
+
+    @Override
+    public long getLength() {
+      if (channel != null && stream == null) {
+        try {
+          channel.close();
+        } catch (Exception e) {
+          throw new IllegalStateException("Attempting to close byte channel", e);
+        }
+        int length = channel.getCount();
+        partitionLengths[partitionId] = length;
+        return length;
+      } else {
+        try {
+          stream.close();
+        } catch (Exception e) {
+          throw new IllegalStateException("Attempting to close output stream", e);
+        }
+        int length = stream.getCount();
+        partitionLengths[partitionId] = length;
+        return length;
+      }
+    }
+
+    @Override
+    public WritableByteChannel openChannel() throws IOException {
+      channel = new PartitionWriterChannel();
+      return channel;
+    }
+  }
+
+  private class PartitionWriterStream extends OutputStream {
+    private int count = 0;
+    private boolean isClosed = false;
+
+    public int getCount() {
+      return count;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+      if (isClosed) {
+        throw new IllegalStateException("Attempting to write to a closed block output stream.");
+      }
+      outputBufferedFileStream.write(b);
+      count++;
+    }
+
+    @Override
+    public void close() throws IOException {
+      flush();
+      isClosed = true;
+    }
+
+    @Override
+    public void flush() throws IOException {
+      outputBufferedFileStream.flush();
+    }
+  }
+
+  private class PartitionWriterChannel implements WritableByteChannel {
+
+    private int count = 0;
+    private boolean isClosed = false;
+
+    public int getCount() {
+      return count;
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+      if (isClosed) {
+        throw new IllegalStateException("Attempting to write to a closed block byte channel.");
+      }
+      int written = outputFileChannel.write(src);
+      count += written;
+      return written;
+    }
+
+    @Override
+    public boolean isOpen() {
+      return !isClosed;
+    }
+
+    @Override
+    public void close() {
```

Comment:
I wonder if we need some equivalent of `flush()` here, especially since the method we use to write to the `WritableByteChannel` is called `Utils.copyFileStreamNIO`, which is non-blocking, so I'm not sure we're guaranteed to have written everything by the time `getCount()` is called. I was trying to do some research on byteChannels and the closest thing I found was `force()`

Reply:
@mccheah any ideas here, I am a bit worried about this being non-blocking as well

Reply:
I looked at `Utils.copyFileStreamNIO` which doesn't have any indication that it's non-blocking, any idea why we think it is?

Reply:
I assumed `NIO` stood for non-blocking IO. The javadoc for the underlying method that it's using, `FileChannel.transferTo()` also implies that the underlying channel could be non-blocking.

Reply:
@mccheah and I just discussed some of this offline.

The `FileChannel.transferTo()` implementation checks for whether the `WriteableByteChannel` is an instance of `FileChannelImpl` and behaves different if it isn't. Because we're not passing in a `FileChannelImpl`, this might actually behave in ways that we don't expect that deviate from the current codepath. Therefore, we shouldn't be wrapping the the `FileChannel` with another `WritableByteChannel` as we're currently doing.

This limits our choices in how we can get the lengths to return in `DefaultShufflePartitionWriter.getLengths()`. The simplest way seems to be to use `FileChannel.position()` calls to get the positions when instantiating the byte channel and when closing the byte channel in `DefaultShufflePartitionWriter.getLengths()`.

Also, looking at the `DefaultShufflePartitionWriter.getLengths()` method, since we want to ensure that the shuffle partitions writers are all closed after getting the lengths, perhaps we should rename this method to `close()` so that it's clear what's should be happening (i.e. you can only get the lengths once everything is written)